### PR TITLE
Allow setting of the output Writer for PRINT and DUMP

### PR DESCRIPTION
**Pull request by andydotxyz**
Requires a concrete Environment type for builtin.
Env was not used for anything so this seemed safe...

### CODE REVIEW

**Comment by andydotxyz** [comment]
I could not see any code using "env interface{}" so I replaced it.
If that functionality needs to be maintained then I could add Environment.Data as interface{}?

**Comment by skx** [comment]
I'm not averse to this, I guess it would be useful to make the library more useful for embedded users?

I guess there are two obvious comments:

* This should also be done for `INPUT`.
* There should be tests.
  * i.e. setup an interpreter, with a fake output, then run `PRINT 3 * 3` and assert the output is 9.

**Comment by andydotxyz** [comment]
Great! I agree on both points. This is already tested - I added the assertion to the PRINT and DUMP tests that existed already - do you want more?

I will look at the INPUT next - I think it should be moved to builtin now that it is possible.

**Comment by skx** [comment]
> I could not see any code using "env interface{}" so I replaced it.

Take a look at `goserver/main.go`.  That uses `env` to invoke allowing state to be maintained between `save` and other code.   It is a contrived example, but without that you can't have any builtins that set/update global state, so it is worth keeping for embedded users.

> If that functionality needs to be maintained then I could add Environment.Data as interface{}?

That works, I think.

**Comment by skx** [comment]
>Great! I agree on both points. This is already tested - I added the assertion to the PRINT and DUMP tests that existed already - do you want more?

No, that's fine.  I read quickly and missed them.  Perfect :)


> I will look at the INPUT next - I think it should be moved to builtin now that it is possible.

Sadly you can't make it a built-in, with the way they work at the moment.  (Ditto READ for the same reason)

If you move the code to a builtin and call:

10 INPUT "Steve", A

The builtin-handler will be called with two arguments:

* The string "Steve"
* The __contents__ of the variable A, if it exists.

What the INPUT handler needs is to be able to be passed a reference to the variable, so it can update the value.  (+/- A vs. A$).

I think I might have documented that at some point, but it's a blocker to move it to builtin, and out of the core.

Anyway I'll check back tomorrow.  Thanks for your contribution (and your fast replies!  Bedtime here now though.)

**Comment by andydotxyz** [comment]
I think that addresses your comments, thanks for the input.
Goserver seems to be working to the best of my knowledge.

File: eval/eval.go [hunk 1]
```diff
@@ -1676,7 +1694,7 @@ func (e *Interpreter) runINPUT() error {
 	var input string
 
 	if flag.Lookup("test.v") == nil {
```
Comment (skx) [inline eval/eval.go]: This code can be improved now; the intent was to allow some minimal testing even though we couldn't fake/stub INPUT-reading.

Your pull-request allows a real test to be made, so the runtime lookup of the test.v flag can go.  

If you update the test-case to cover this then I think everything looks good :)

**Comment by skx** [comment]
Merged, thanks again for your contribution :)

**Pull request merged**